pyboolnet/control_strategies.py

```diff
@@ -0,0 +1,486 @@
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *cs*: list of control strategies.
+        * *target*: list of subspaces defining the target subset.
+
+    **returns**:
+        * *cs_perc*: list of control strategies by direct percolation.
+    """
+
+    cs_perc = list()
+    for x in cs:
+        perc = find_constants(primes=percolate(primes=primes, add_constants=x, copy=True))
+        if any(is_included_in_subspace(perc, subs) for subs in target):
+            cs_perc.append(x)
+    return cs_perc
+
+
+def control_is_valid_in_trap_spaces(primes, trap_spaces, target, update):
+    """
+    Returns wheter the *trap_spaces* are compatible with the *target*.
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *trap_spaces*: list of trap spaces.
+        * *target*: list of subspaces defining the target subset.
+        * *update*: type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+
+    **returns**:
+        * *cs_perc*: list of control strategies by direct percolation.
+    """
+
+    # Checking no trap spaces are disjoint from the target subset
+    if not all(any(dicts_are_consistent(ts, subs) for subs in target) for ts in trap_spaces):
+        return False
+
+    # Checking trap spaces oscillating in-out the target subset
+    half_ts = [ts for ts in trap_spaces if not any(is_included_in_subspace(ts, subs) for subs in target)]
+    for ts in half_ts:
+        if not reduce_and_run_control_query(primes, ts, target, update):
+            return False
+
+    return True
+
+
+def reduce_and_run_control_query(primes, subspace, target, update):
+    """
+    Runs the model checking query for control after reducing the network by percolating the values in *subspace*.
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *subspace*: subspace.
+        * *target*: list of subspaces defining the target subset.
+        * *update*: type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+
+    **returns**:
+        * True if the query is true, false otherwise.
+    """
+
+    target_vars = list(set([item for subs in target for item in subs]))
+    new_primes = fix_components_and_reduce(primes, subspace, keep_vars=target_vars)
+    return run_control_query(new_primes, target, update)
+
+
+def run_control_query(primes, target, update):
+    """
+    Runs the model checking query for control.
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *target*: list of subspaces defining the target subset.
+        * *update*: type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+
+    **returns**:
+        * True if the query is true, False otherwise.
+    """
+
+    spec = 'CTLSPEC ' + EFAG_set_of_subspaces(primes, target)
+    init = "INIT TRUE"
+    return model_checking(primes, update, init, spec)
+
+
+def control_direct_percolation(primes, candidate, target, silent=False):
```

Comment:
print control via a parameter `silent` is outdated. Instead we use a logger now. The user can control what info to see by setting log levels before running scripts, see one of the other modules. The key lines are:

```
import logging
log = logging.getLogger(__name__)
```

and instead of print you can write
```
log.info("this info is printed at level 'info'")
log.debug("this info is printed at level 'debug'")
```